File: libs/wasm-sdk/wasm-sdk-kotlin/src/wasmWasiMain/kotlin/sdk/Host.kt
package sdk

// ============================================================================
// flowlike_log
// ============================================================================

@WasmImport("flowlike_log", "debug")
private external fun hostLogDebug(ptr: Int, len: Int)

@WasmImport("flowlike_log", "info")
private external fun hostLogInfo(ptr: Int, len: Int)

@WasmImport("flowlike_log", "warn")
private external fun hostLogWarn(ptr: Int, len: Int)

@WasmImport("flowlike_log", "error")
private external fun hostLogError(ptr: Int, len: Int)

@WasmImport("flowlike_log", "trace")
private external fun hostLogTrace(ptr: Int, len: Int)

@WasmImport("flowlike_log", "log_json")
private external fun hostLogJson(level: Int, msgPtr: Int, msgLen: Int, dataPtr: Int, dataLen: Int)

// ============================================================================
// flowlike_pins
// ============================================================================

@WasmImport("flowlike_pins", "get_input")
private external fun hostGetInput(namePtr: Int, nameLen: Int): Long

@WasmImport("flowlike_pins", "set_output")
private external fun hostSetOutput(namePtr: Int, nameLen: Int, valuePtr: Int, valueLen: Int)

@WasmImport("flowlike_pins", "activate_exec")
private external fun hostActivateExec(namePtr: Int, nameLen: Int)

// ============================================================================
// flowlike_vars
// ============================================================================

@WasmImport("flowlike_vars", "get")
private external fun hostVarGet(namePtr: Int, nameLen: Int): Long

@WasmImport("flowlike_vars", "set")
private external fun hostVarSet(namePtr: Int, nameLen: Int, valuePtr: Int, valueLen: Int)

@WasmImport("flowlike_vars", "delete")
private external fun hostVarDelete(namePtr: Int, nameLen: Int)

@WasmImport("flowlike_vars", "has")
private external fun hostVarHas(namePtr: Int, nameLen: Int): Int

// ============================================================================
// flowlike_cache
// ============================================================================

@WasmImport("flowlike_cache", "get")
private external fun hostCacheGet(keyPtr: Int, keyLen: Int): Long

@WasmImport("flowlike_cache", "set")
private external fun hostCacheSet(keyPtr: Int, keyLen: Int, valPtr: Int, valLen: Int)

@WasmImport("flowlike_cache", "delete")
private external fun hostCacheDelete(keyPtr: Int, keyLen: Int)

@WasmImport("flowlike_cache", "has")
private external fun hostCacheHas(keyPtr: Int, keyLen: Int): Int

// ============================================================================
// flowlike_meta
// ============================================================================

@WasmImport("flowlike_meta", "get_node_id")
private external fun hostGetNodeId(): Long

@WasmImport("flowlike_meta", "get_run_id")
private external fun hostGetRunId(): Long

@WasmImport("flowlike_meta", "get_app_id")
private external fun hostGetAppId(): Long

@WasmImport("flowlike_meta", "get_board_id")
private external fun hostGetBoardId(): Long

@WasmImport("flowlike_meta", "get_user_id")
private external fun hostGetUserId(): Long

@WasmImport("flowlike_meta", "is_streaming")
private external fun hostIsStreaming(): Int

@WasmImport("flowlike_meta", "get_log_level")
private external fun hostGetLogLevel(): Int

@WasmImport("flowlike_meta", "time_now")
private external fun hostTimeNow(): Long

@WasmImport("flowlike_meta", "random")
private external fun hostRandom(): Long

// ============================================================================
// flowlike_stream
// ============================================================================

@WasmImport("flowlike_stream", "emit")
private external fun hostStreamEmit(eventPtr: Int, eventLen: Int, dataPtr: Int, dataLen: Int)

@WasmImport("flowlike_stream", "text")
private external fun hostStreamText(textPtr: Int, textLen: Int)

// ============================================================================
// flowlike_storage
// ============================================================================

@WasmImport("flowlike_storage", "read_request")
private external fun hostStorageRead(pathPtr: Int, pathLen: Int): Long

@WasmImport("flowlike_storage", "write_request")
private external fun hostStorageWrite(pathPtr: Int, pathLen: Int, dataPtr: Int, dataLen: Int): Int

@WasmImport("flowlike_storage", "storage_dir")
private external fun hostStorageDir(nodeScoped: Int): Long

@WasmImport("flowlike_storage", "upload_dir")
private external fun hostUploadDir(): Long

@WasmImport("flowlike_storage", "cache_dir")
private external fun hostCacheDir(nodeScoped: Int, userScoped: Int): Long

@WasmImport("flowlike_storage", "user_dir")
private external fun hostUserDir(nodeScoped: Int): Long

@WasmImport("flowlike_storage", "list_request")
private external fun hostStorageList(pathPtr: Int, pathLen: Int): Long

// ============================================================================
// flowlike_models
// ============================================================================

@WasmImport("flowlike_models", "embed_text")
private external fun hostEmbedText(bitPtr: Int, bitLen: Int, textsPtr: Int, textsLen: Int): Long

// ============================================================================
// flowlike_http
// ============================================================================

@WasmImport("flowlike_http", "request")
private external fun hostHttpRequest(
    method: Int,
    urlPtr: Int, urlLen: Int,
    headersPtr: Int, headersLen: Int,
    bodyPtr: Int, bodyLen: Int,
): Int

// ============================================================================
// flowlike_auth
// ============================================================================

@WasmImport("flowlike_auth", "get_oauth_token")
private external fun hostGetOauthToken(providerPtr: Int, providerLen: Int): Long

@WasmImport("flowlike_auth", "has_oauth_token")
private external fun hostHasOauthToken(providerPtr: Int, providerLen: Int): Int

// ============================================================================
// Wrapper functions
// ============================================================================

// -- Logging --

fun logDebug(message: String) {
    val (ptr, len) = stringToPtr(message)
    hostLogDebug(ptr, len)
}

fun logInfo(message: String) {
    val (ptr, len) = stringToPtr(message)
    hostLogInfo(ptr, len)
}

fun logWarn(message: String) {
    val (ptr, len) = stringToPtr(message)
    hostLogWarn(ptr, len)
}

fun logError(message: String) {
    val (ptr, len) = stringToPtr(message)
    hostLogError(ptr, len)
}

fun logTrace(message: String) {
    val (ptr, len) = stringToPtr(message)
    hostLogTrace(ptr, len)
}

fun logJson(level: Int, message: String, data: String) {
    val (msgPtr, msgLen) = stringToPtr(message)
    val (dataPtr, dataLen) = stringToPtr(data)
    hostLogJson(level, msgPtr, msgLen, dataPtr, dataLen)
}

// -- Pins --

fun getInput(name: String): String? {
    val (ptr, len) = stringToPtr(name)
    val result = hostGetInput(ptr, len)
    return unpackString(result)
}

fun setOutput(name: String, value: String) {
    val (namePtr, nameLen) = stringToPtr(name)
    val (valPtr, valLen) = stringToPtr(value)
    hostSetOutput(namePtr, nameLen, valPtr, valLen)
}

fun activateExec(name: String) {
    val (ptr, len) = stringToPtr(name)
    hostActivateExec(ptr, len)
}

// -- Variables --

fun getVariable(name: String): String? {
    val (ptr, len) = stringToPtr(name)
    return unpackString(hostVarGet(ptr, len))
}

fun setVariable(name: String, value: String) {
    val (namePtr, nameLen) = stringToPtr(name)
    val (valPtr, valLen) = stringToPtr(value)
    hostVarSet(namePtr, nameLen, valPtr, valLen)
}

fun deleteVariable(name: String) {
    val (ptr, len) = stringToPtr(name)
    hostVarDelete(ptr, len)
}

fun hasVariable(name: String): Boolean {
    val (ptr, len) = stringToPtr(name)
    return hostVarHas(ptr, len) != 0
}

// -- Cache --

fun cacheGet(key: String): String? {
    val (ptr, len) = stringToPtr(key)
    return unpackString(hostCacheGet(ptr, len))
}

fun cacheSet(key: String, value: String) {
    val (keyPtr, keyLen) = stringToPtr(key)
    val (valPtr, valLen) = stringToPtr(value)
    hostCacheSet(keyPtr, keyLen, valPtr, valLen)
}

fun cacheDelete(key: String) {
    val (ptr, len) = stringToPtr(key)
    hostCacheDelete(ptr, len)
}

fun cacheHas(key: String): Boolean {
    val (ptr, len) = stringToPtr(key)
    return hostCacheHas(ptr, len) != 0
}

// -- Stream --

fun stream(eventType: String, data: String) {
    val (evtPtr, evtLen) = stringToPtr(eventType)
    val (dataPtr, dataLen) = stringToPtr(data)
    hostStreamEmit(evtPtr, evtLen, dataPtr, dataLen)
}

fun streamTextRaw(text: String) {
    val (ptr, len) = stringToPtr(text)
    hostStreamText(ptr, len)
}

fun streamText(text: String) {
    stream("text", text)
}

// -- Meta --

fun getNodeId(): String? = unpackString(hostGetNodeId())
fun getRunId(): String? = unpackString(hostGetRunId())
fun getAppId(): String? = unpackString(hostGetAppId())
fun getBoardId(): String? = unpackString(hostGetBoardId())
fun getUserId(): String? = unpackString(hostGetUserId())
fun isStreaming(): Boolean = hostIsStreaming() != 0
fun getLogLevel(): Int = hostGetLogLevel()
fun timeNow(): Long = hostTimeNow()
fun random(): Long = hostRandom()

// -- Storage --

fun storageRead(path: String): String? {
    val (ptr, len) = stringToPtr(path)
    return unpackString(hostStorageRead(ptr, len))
}

fun storageWrite(path: String, data: String): Boolean {
    val (pathPtr, pathLen) = stringToPtr(path)
    val (dataPtr, dataLen) = stringToPtr(data)
    return hostStorageWrite(pathPtr, pathLen, dataPtr, dataLen) != 0
}

fun storageDir(nodeScoped: Boolean): String? =
    unpackString(hostStorageDir(if (nodeScoped) 1 else 0))

fun uploadDir(): String? = unpackString(hostUploadDir())

fun cacheDir(nodeScoped: Boolean, userScoped: Boolean): String? =
    unpackString(hostCacheDir(if (nodeScoped) 1 else 0, if (userScoped) 1 else 0))

fun userDir(nodeScoped: Boolean): String? =
    unpackString(hostUserDir(if (nodeScoped) 1 else 0))

fun storageList(path: String): String? {
    val (ptr, len) = stringToPtr(path)
    return unpackString(hostStorageList(ptr, len))
}

File: libs/wasm-sdk/wasm-sdk-swift/Sources/FlowLikeSDK/Host.swift
// Host.swift — WASM host import bindings for all Flow-Like modules.
//
// All host functions are declared in the FlowLikeHostC C target using
// __attribute__((import_module, import_name)) so that they emit clean
// C-ABI WASM imports (no hidden Swift context/error parameters).

import FlowLikeHostC

// MARK: - Logging

public func logTrace(_ msg: String) {
    let (p, l) = stringToPtr(msg)
    flowlike_log_trace(p, l)
}

public func logDebug(_ msg: String) {
    let (p, l) = stringToPtr(msg)
    flowlike_log_debug(p, l)
}

public func logInfo(_ msg: String) {
    let (p, l) = stringToPtr(msg)
    flowlike_log_info(p, l)
}

public func logWarn(_ msg: String) {
    let (p, l) = stringToPtr(msg)
    flowlike_log_warn(p, l)
}

public func logError(_ msg: String) {
    let (p, l) = stringToPtr(msg)
    flowlike_log_error(p, l)
}

public func logJSON(level: Int, msg: String, data: String) {
    let (mp, ml) = stringToPtr(msg)
    let (dp, dl) = stringToPtr(data)
    flowlike_log_json(Int32(level), mp, ml, dp, dl)
}

// MARK: - Pins

public func getInput(_ name: String) -> String {
    let (p, l) = stringToPtr(name)
    return unpackString(flowlike_pins_get_input(p, l))
}

public func setOutput(_ name: String, _ value: String) {
    let (np, nl) = stringToPtr(name)
    let (vp, vl) = stringToPtr(value)
    flowlike_pins_set_output(np, nl, vp, vl)
}

public func activateExec(_ name: String) {
    let (p, l) = stringToPtr(name)
    flowlike_pins_activate_exec(p, l)
}

// MARK: - Variables

public func getVariable(_ name: String) -> String {
    let (p, l) = stringToPtr(name)
    return unpackString(flowlike_vars_get(p, l))
}

public func setVariable(_ name: String, _ value: String) {
    let (np, nl) = stringToPtr(name)
    let (vp, vl) = stringToPtr(value)
    flowlike_vars_set(np, nl, vp, vl)
}

public func deleteVariable(_ name: String) {
    let (p, l) = stringToPtr(name)
    flowlike_vars_delete(p, l)
}

public func hasVariable(_ name: String) -> Bool {
    let (p, l) = stringToPtr(name)
    return flowlike_vars_has(p, l) != 0
}

// MARK: - Cache

public func cacheGet(_ key: String) -> String {
    let (p, l) = stringToPtr(key)
    return unpackString(flowlike_cache_get(p, l))
}

public func cacheSet(_ key: String, _ value: String) {
    let (kp, kl) = stringToPtr(key)
    let (vp, vl) = stringToPtr(value)
    flowlike_cache_set(kp, kl, vp, vl)
}

public func cacheDelete(_ key: String) {
    let (p, l) = stringToPtr(key)
    flowlike_cache_delete(p, l)
}

public func cacheHas(_ key: String) -> Bool {
    let (p, l) = stringToPtr(key)
    return flowlike_cache_has(p, l) != 0
}

// MARK: - Metadata

public func getNodeID() -> String { unpackString(flowlike_meta_get_node_id()) }
public func getRunID() -> String { unpackString(flowlike_meta_get_run_id()) }
public func getAppID() -> String { unpackString(flowlike_meta_get_app_id()) }
public func getBoardID() -> String { unpackString(flowlike_meta_get_board_id()) }
public func getUserID() -> String { unpackString(flowlike_meta_get_user_id()) }
public func isStreaming() -> Bool { flowlike_meta_is_streaming() != 0 }
public func getLogLevelValue() -> Int { Int(flowlike_meta_get_log_level()) }
public func timeNow() -> Int64 { flowlike_meta_time_now() }
public func random() -> Int64 { flowlike_meta_random() }

// MARK: - Storage

public func storageRead(_ path: String) -> String {
    let (p, l) = stringToPtr(path)
    return unpackString(flowlike_storage_read_request(p, l))
}

public func storageWrite(_ path: String, _ data: String) -> Bool {
    let (pp, pl) = stringToPtr(path)
    let (dp, dl) = stringToPtr(data)
    return flowlike_storage_write_request(pp, pl, dp, dl) != 0
}

public func storageDir(nodeScoped: Bool) -> String {
    unpackString(flowlike_storage_storage_dir(nodeScoped ? 1 : 0))
}

public func uploadDir() -> String {
    unpackString(flowlike_storage_upload_dir())
}

public func cacheDirPath(nodeScoped: Bool, userScoped: Bool) -> String {
    unpackString(flowlike_storage_cache_dir(nodeScoped ? 1 : 0, userScoped ? 1 : 0))
}

public func userDir(nodeScoped: Bool) -> String {
    unpackString(flowlike_storage_user_dir(nodeScoped ? 1 : 0))
}

public func storageList(_ flowPathJSON: String) -> String {
    let (p, l) = stringToPtr(flowPathJSON)
    return unpackString(flowlike_storage_list_request(p, l))
}

// MARK: - Models

public func embedText(bitJSON: String, textsJSON: String) -> String {
    let (bp, bl) = stringToPtr(bitJSON)
    let (tp, tl) = stringToPtr(textsJSON)
    return unpackString(flowlike_models_embed_text(bp, bl, tp, tl))
}

// MARK: - HTTP

public func httpRequest(method: Int, url: String, headers: String, body: String) -> Bool {
    let (up, ul) = stringToPtr(url)
    let (hp, hl) = stringToPtr(headers)
    let (bp, bl) = stringToPtr(body)
    return flowlike_http_request(Int32(method), up, ul, hp, hl, bp, bl) != 0
}

// MARK: - Streaming

public func streamEmit(eventType: String, data: String) {
    let (ep, el) = stringToPtr(eventType)
    let (dp, dl) = stringToPtr(data)
    flowlike_stream_emit(ep, el, dp, dl)
}

public func streamText(_ text: String) {
    let (p, l) = stringToPtr(text)
    flowlike_stream_text(p, l)
}

// MARK: - Auth

public func getOAuthToken(_ provider: String) -> String {
    let (p, l) = stringToPtr(provider)
    return unpackString(flowlike_auth_get_oauth_token(p, l))
}

public func hasOAuthToken(_ provider: String) -> Bool {
    let (p, l) = stringToPtr(provider)
    return flowlike_auth_has_oauth_token(p, l) != 0
}

